minctracc/minctracc.1
.\" Copyright 1995 Louis Collins, McConnell Brain Imaging Centre,
.\" Montreal Neurological Institute, McGill University.
.\" Permission to use, copy, modify, and distribute this
.\" software and its documentation for any purpose and without
.\" fee is hereby granted, provided that the above copyright
.\" notice appear in all copies.  The author and McGill University
.\" make no representations about the suitability of this
.\" software for any purpose.  It is provided "as is" without
.\" express or implied warranty.
.\"
.\" $Header: /private-cvsroot/registration/mni_autoreg/minctracc/minctracc.1,v 1.3 2011-11-08 02:50:11 rotor Exp $
.\"
.TH MINCTRACC 1

.SH NAME
minctracc - estimates the spatial transformation required to register two volumes together.

.SH SYNOPSIS
.B minctracc [<options>] <source> <target> <output>

.B minctracc [-help]


.SH DESCRIPTION
.I Minctracc
will estimate the best linear spatial transformation required to
register two 3D volumes.  The program uses optimization over a user
selectable number of parameters to identify the best (according to a
user identified objective function) transformation that maps voxel
values of the first data set into the second.  The objective function
is evaluated on a 3D lattice with user-specified spacing. 

.SH OVERVIEW 

You need to specify two MINC data files on the command line, the first
is the source volume and the second is the target volume.  The
third item on the command line is the name of the output
transformation file, specifying where the program will store the
results (in .xfm format).  Note that the output is 
.I not
a new MINC volume; you must use mincresample to resample the data
according to the transform output by minctracc if you wish to see the
transformed image data.

Using flags on the command line, the user can specify:

 - the type of transformation to be optimized and
   saved, 
 - mask files to be used with the source or target 
   volumes, 
 - a transformation file to specify the initial 
   starting condition, 
 - the type of objective function to be minimized, 
 - the type of interpolation function to use, 
 - and the definition of the sampling matrix used 
   to measure the objective function.

Internally, the program optimizes the transformation parameters of
translations, rotations, scaling(s) and possibly shear(s).  These
parameters are used to create a homogeneous transformation matrix that
maps the data in the world space of the first volume to the world
space of the second volume.  It is this matrix that is saved in the
output transformation file.  The program 'xfm2param' can be used to
extract the transformation parameters from the .xfm file.

.SH Specifying the initial starting position:

There are a number of methods available to get an initial
transformation to start the optimization process.  The first is the
default if there is no input transformation information on the command
line.  In this case, a principal axes transformation (PAT) is used as
the starting transformation.  When the PAT is used, the center of
rotation and scaling is automatically estimated from the center of
gravity (COG) of the source data set.  The initial translation
parameters are those needed to align the COGs of the two data sets.
The rotations are those need to align the principal axes of the two
volumes.  This method is not always robust when the two data sets do
not represent the same volume (ie source is 10cm thick slice of brain,
and the target is a whole head), so other initialization methods are
available.

In the second method, with the '-transformation' option the user
specifies a transform (.xfm) file that contains a linear
transformation used to map points of volume one (source) into volume
two (target).  This transformation file can be the result of a rough
manual registration accomplished with 'register', or the previous
result of an application of minctracc, for example.

The third option is used when the two volumes are already fairly
closely aligned (eg., if they were acquired in the same scanning
session).  Here, '-identity' is used to specify an identity matrix to
the used to initialize the transformation matrix.

Once the matrix is specified, the program extracts the transformation
parameters (translations, rotations, scales and shears) and optimizes
these parameters when trying to find the best transformation.

When the user specifies a transformation matrix (other than PAT), the
default center is 0,0,0.  The user can can change this in two ways: 1)
the '-center <cx> <cy> <cz>' option permits direct input of the center
to be used, 2) the '-est_center' option tells the program to estimate
and use the center of gravity of the source volume.  Note that it is a
good idea to have a reasonable guess for the center of rotation and
scaling since it makes the parameters more orthogonal to each other
(e.g.  a small change in rotation will not grossly affect the
translations).

Note that when estimating a non-linear registration, if no
transformation is given as input, the program will estimate a linear
transformation first.  If you are interested in only the non-linear
transformation, then use -identity.

.SH OPTIONS
Note that options can be specified in abbreviated form (as long as
they are unique) and can be given anywhere on the command line.

.SH Initial transformation information.
These are the options used to specify the initial starting conditions
for the optimization of the transformation parameters.
.P
.I -transformation
<filename.xfm>:
Specify a file giving the world coordinate transformation mapping
points from the source into the target space.
.P
.I -identity:
Use identity transformation for starting point.
.P
.I -est_center:
Use only the center estimated from Principal axis trans.
.P
.I -est_scales:
Use only the scales estimated from Principal axis trans.
.\" .P
.\" .I -est_rotations:
.\" Use only the rotations estimated from Principal axis trans.
.P
.I -est_translations:
Use only the translations estimated from Principal axis trans.
.P
Note that the four previous options (-est_*) are not mutually exclusive. You
can, for example, use '-est_center -est_translations' to estimate the
center of rotation and scaling, as well as the initial translation
componant required to register the centers of gravity of the two
volumes.
.P
.I -center
<xcent> <ycent> <zcent>: Force the center of rotation and scale.
.P
.I -no_clobber:
Do not overwrite output file (default).
.P
.I -clobber:
Overwrite output file.
.SH Output transformation type. 
The type of output file can be constrained depending on the type of
registration required.  For example, intra-subject registration can be
completed with only 6 parameters if rigid body assumptions hold.
.P
.I -pat:
Return the principal axis transformation matrix (input matrix
ignored), no optimization done.
.P
.I -lsq3:
Return a 3 parameter transformation (3 translations only).
.P
.I -lsq6:
Return a 6 parameter rigid-body transformation (3 translation, 3 rotation, scale=1.0).
.P
.I -lsq7:
Return a 7 parameter transformation (lsq6 + 1 global scale; this is
the same as -procrustes, and is the default).
.P
.I -lsq9:
Return a 9 parameter transformation (lsq6 + 3 scales).
.P
.I -lsq10:
Return a 10 parameter transformation (lsq9 + 1 shear).
.P
.I -lsq12:
Return a full 12 parameter transformation (lsq9 + 3 shears).
.P
.I 
-lsq:
Return a full 12 parameter transformation (same as -lsq12).
.P
.I -procrustes:
Return a procrustes transformation (3 trans, 3 rots, 1 scale), same as
-lsq7.  This is the default.
.P
.I -forward:
Recover forward transformation (source -> model, default).
.P
.I -invert:
Recover inverted transformation (model -> source).
.SH Options for mask volumes.
.P
.I -model_mask
<filename>:
Specifies a binary mask file for the target.  Any data voxel whose
world coordinate falls in a zero-valued voxel in the mask is ignored
in the calculation of the objective function.
.P
.I -source_mask
<filename>:
Specifies a binary mask file for the source.
.SH Interpolation options.
.P
.I -trilinear:
Do a tri-linear interpolation between voxels when estimating the value
for a node in the 3D lattice.  This is the default.
.P
.I -tricubic:
Do a tri-cubic interpolation between voxels.
.P
.I -nearest_neighbour:
Do nearest neighbour interpolation between voxels (ie. find the voxel
closest to the point and use its value). 
.SH Optimization objective functions. 
.P
.I -xcorr:
Use cross correlation (this is the default) [2].
.P
.I -zscore:
Use normalized difference.  Before optimization, each volume is
normalized to have the same mean, with a range of +/- 5 standard
deviations.  The objective function is simply the difference in
normalized values between the two volumes.
.P
.I -ssc:
Use stochastic sign change [3].  This is the same as
maximization of zero crossings. 
.P
.I -vr:
Minimize the variance of the ratio vol1/vol2 [4].
.P
.I -mi:
Use mutual information similarity measure [1].
.P
.I -groups
<num>: Number of groups for -vr and -mi (default =  16).
.P
.I -threshold
<thresh1> <thresh2>: Lower limit for voxel threshold (default = 0.0
0.0).  Lattice nodes with voxel values below this limit are ignored in
the objective function. The threshold applies when -xcorr, -vr, -mi or
-zscore are used as the objective function.
.P
.I -speckle
<val>: Percent speckle noise to add to source (default = 5). This option applies only when -ssc is used as the objective function.


.SH Options for linear optimization.
Simplex optimization is used to maximize/minimize the objective
function to find the best transformation parameters.
.P
.I -tol
<val>: Stopping criteria tolerance (default = 0.005).  The
optimization will stop when tol> (f_max-f_min)/(f_max+f_min), where
f_max and f_min are the maximum and minimum values of the objective
function in the Simplex.
.P
.I -simplex
<val>: Radius of simplex volume (default = 20).  This is measured in
units of millimeters on the spatial axis, degrees of rotation or
percentage scale on the scaling dimensions.  When the initial
estimate is know to be relatively good, the simplex radius should be
reduced to the level of certainty of the input parameters.
.P
.I -w_translations
<w_tx> <w_ty> <w_tz>: Optimization weight of translation in x, y, z
(default = 1.0 1.0 1.0).
.P
.I -w_rotations
<w_rx> <w_ry> <w_rz>: Optimization weight of rotations around x, y, z
(default = 0.0174533 0.0174533 0.0174533).  Internally, the rotations
are stored as radians, although all user input is in degrees.  The
value 0.0174533 makes one degree equivalent to 1 mm in the optimization.
.P
.I -w_scales
<w_sx> <w_sy> <w_sz>: Optimization weight of scaling along x, y, z
(default = 0.02 0.02 0.02).  This makes a 2% change in scale
equivalent to 1mm of translation.
.P
.I -w_shear
<w_sa> <w_sb> <w_sc>: Optimization weight of shears a,b and c (default
= 0.02 0.02 0.02)
.SH Options for 3D lattice definition.
The objective function is estimated only on the nodes of a 3D lattice
defined on the smallest of the two volumes.  In this way, the
coordinates of the lattice are used to specify positions in one
volume, and when mapped through the transformation matrix, specify
homologous positions in the other volume.  The spacing between lattice
samples is directly related to the resolution of the data used in the
fit.  For example, data blurred with a 12mm fwhm gaussian kernel does
not need to be sampled with spacing less than 6mm.
.P
.I -step
<sx> <sy> <sz>: Step size (in mm) along each dimension (x, y, z).  Default
value is 4.0 4.0 4.0.
.P
.I -xstep
<sx>: Step size along the X dimension (default = 4.0).
.P
.I -ystep
<sy>: Step size along the Y dimension (default = 4.0).
.P
.I -zstep
<sz>: Step size along the Z dimension (default = 4.0)
.SH Options for optimization of non-linear transformations

The non-linear transformation is represented by a deformation field,
(normally) defined in the space of the target volume.  The full
transformation is equal to the linear transformation plus the
deformation stored in the deformation field.  The deformation field is
represented by a vector-valued 3D volume with -step distance between
nodes.

.P
.I -nonlinear
specifies that a non-linear deformation field should be estimated.
the -nonlinear can take on one of the following optional arguments:
xcorr, diff, sqdiff, label, chamfer, corrcoeff, or opticalflow to
define the objective function to be used to compare the source and
target volumes.

.P
.I -sub_lattice
<val>: 
Defines the number of nodes along the diameter of the
sublattice that defined the local neighbourhood used to estimate the
deformation vector.
.P
.I -lattice_diameter
<valx> <valy> <valz>:
determines the size (in mm) of the sublattice used to define the local neighbouhood.
.P
.I  -use_magnitude         
use magnitude data local deformation (default).  this flag tells
minctracc that a local sublattice will be needed.  You normally don't
have to specify this flag.
.P
.I   -optical_flow
a flag to use optical flow to compute deformation on the two firstmain source/model volumes.
.P
.I  -use_simplex
a flag to use 3D simplex optimization for local deformation (default).
.P
.I   -quadratic
a flag to turn on local quadratic fitting for local deformation.
.P
.I   -use_local
a flag to turn on local smoothing.  by default, minctracc uses global smoothing for regularization.
.P
.I   -use_nonisotropic
a flag to turn on directionally sensitive smoothing (def=isotropic smoothing) when usinglocal smoothing.
.P
.I   -super
<val>:
super sample deformation field during optimization (default=2) to speed up the evaluation of local deformation vectors.
.P
.I   -no_super
turn off the super sample deformation field during optimization.
.P
.I   -iterations
<val>
this is the number of iterations for non-linear optimization (default value: 4).
.P
.I   -weight
<val>: 
Weighting factor for each iteration in nl optimization. This defines
how much of the currently estimated vector should be added to the
deformation field (default value: 0.6).
.P
.I   -stiffness 
<val>:
Weighting factor to define smoothness for regularization at each iteration (default value: 0.5).
.P
.I   -similarity_cost_ratio
<val>
Weighting factor to reduce the effect of large deformations [ r=similarity*w + cost(1*w) ] (default value: 0.5)

.SH Options for logging progress.
.P
.I -verbose
<val>:
Write verbose messages indicating progress (default = 1).
.P
.I -quiet:
Do not write log messages
.P
.I -debug:
Print out debug info.
.SH Generic options
.P
.I -help:
Print summary of command-line options and abort.

.SH EXAMPLES

Estimate the transformation required to map structures from an 
individual subject to match those in a target volume:

   minctracc subject.mnc target.mnc subj_to_targ.xfm 

Match the same subject, scanned on two occasions with similar
protocol: 

   minctracc subj_time1.mnc subj_time2.mnc result.xfm \\
	-lsq6 -identity -est_center


.SH REFERENCES

[1] A. Collingnon, F. Maes, D. Delaere, D. Vandermeulen, P. Suetens,
G. Marchal, "Automated multi-modality image registration based on
information theory", IPMI, 1995:263-274.

[2] Collins DL, Neelin P, Peters TM, and Evans AC, "Automatic 3-D
intersubject registration of MR volumetric data in standardized
Talairach space", J. Comput. Assis. Tomogr, 18(2):192-205.

[3] Minoshima S, Berger KL, Lee Ks, Mintun MA.  "An automated method
for rotation correction and centering of 3D functional brain images",
J. Nucl. Med., 1992;33(8):1579-1585.

[4] R.P. Woods, J.C. Mazziotta, S.R. Cherry, "MRI-PET Registration
with Automated Algorithm", J. Comput. Assis. Tomogr, 1993;17:536-546.


.SH AUTHOR
Louis Collins

.SH COPYRIGHTS
.ps 18
Copyright 1993-95 by Louis Collins


NEWS
Release 
--------------


Release 0.99.6
--------------
* Free memory after usage in mincblur.
* fixed one bug in apodize_data.c for ABS(step) in one direction
* now use VIO_X, VIO_Y, VIO_Z (no more X, Y, Z) in mincblur and
  make it invariant to dimension order

Release 0.99.5
--------------

* input source and volumes are read as DOUBLES.  
* deformation fields are created and manipulated as DOUBLES.
	
Release 0.99.4
--------------

* added -nmi option (normalized mutual information) as an objective
  function for linear registration.
* modeldir is now $prefix/share/mni-models
* fixed a bug for offsets in minctracc/Volume/init_lattice.c
* free memory after usage in minctracc
* file compression (minc2) disabled for temporary files in make_model
* fixed bug in mritoself with same file names in different directories

Release 0.99.3
--------------

* new release to compile on ia64

Release 0.99.2
--------------

* fixed a bug in mritoself and mritotal when piping output to STDOUT

Release 0.99.1
--------------

* fixed a bug in autocrop regarding split_path

Release 0.98v
-------------

* added vladimir fonovs changes to xfmtool to allow .gz grid files


Release 0.98u
-------------

* removed perl libraries that are now part of mni_perllib
* removed c libraries that are now part of MINC proper

Release 0.98t
-------------

* 30% speed increase due to changes in nonlinear fitting code
* Corrected masking in sub-lattice (Patricia Le Nezet and Irina Kezele)


Release 0.98s
-------------

* Removed malloc.h from mincblur to assist build on OSX


Release 0.98r
-------------

* Fixed COV bug (lenezet)
* Fixed make check bug with direction cosines
* added epm-header.in for binary package building


Release 0.98q
-------------

* Removed all reference to public and private syntatic sugar
    private has been replaced with static.  This has caused
	erratic build problems in the past.  Let's see if things
	are better now.
* Removed most Keyword Substitutions in perl programs in favour
    of the perl module FindBin::Bin this module is part of the
	base perl distribution


Release 0.98p
-------------

* Fixed problems with Direction cosines of input volumes
* Changed from volume_stats to mincstats in minc_utilities.pl
   this will mean that we no longer have a dependency on N3.Z


Release 0.98j
-------------

* Build process changed: third-party libraries will no longer be
  searched for in /usr/local and ${prefix} by default.  Specify
  the directories to search using --with-build-path.


Release 0.98i
-------------

This is a bug-fix release; no user-visible changes.


Release 0.98h
-------------

* mritotal handles nonlinear as well as linear fitting.
* Build process revamped: supports non source-directory builds; no
  environment variables are required to specify library locations.
* Fixed bug that caused PROCRUSTES transformations to optimize only
  the x-scale, rather than a global scale (as LSQ7 does).
* Generated files contain the long version string in their comment section.


Release 0.98g
-------------

* Minctracc is built with both linear and non-linear registration.


Release 0.98f
-------------

* version 0.98e did not give the same registration results as 0.98d when
  using masks due to an oversight in the voxel coordinate manipulation
  in the mask volumes - an important problem for using mritotal.  
  This has been corrected in 0.98f.


Release 0.98e
-------------

* feature volumes are supported to be able to register sets (like T1, T2 and
  classified data) of volumes from a source data set to match corresponding
  features of a target.
* the minctracc registration engine is approximately 2.5 times faster for
  resampling data during evaluation of the optimization function.  This
  results in an overall reduction in execution time by a factor of 2.


Release 0.98
------------

* added mritoself to list of programs to install (in perl/Makefile)
* slight tweak to test of compiler's ANSIness in configure script
* fixed some subtle but important bugs in mritotal, autocrop, and mritoself
* more enhancements to the Perl libraries
* removed an unused option (-Lvv) from mincblur's argument table
* removed ParseArgs module from the distribution -- Getopt::Tabular 
  (available separately) is now required
* changed mritotal's algorithm for trimming volumes after blurring
* added '-bbox' feature to autocrop
bug fixes:
release 0.98b, 0.98c: fixed two bugs in mritoself


Release 0.97
------------

* fixed those bugs in mritotal
* enhanced configure script for portability and flexibility
* added the TESTING document to explain testing the package in detail


Release 0.96
------------

* inexplicable jump of four micro-versions instead of one
* introduced some buggy features in mritotal


Release 0.92
------------

* the implementation of the mutual information objective function in
  minctracc has been greatly improved
* many portability problems have been fixed
* some real bugs (particularly in mincblur) have been fixed
* most of the Perl libraries have been greatly updated to take
  better advantage of Perl 5
* added -blur/-noblur and -firstobj options to mritotal (experimental --
  not used by any of the standard fitting procedures)


Release 0.91
------------

Lost in the mists of time...


Release 0.9
-----------

First public release.


minctracc/Main/minctracc.c
/* ----------------------------- MNI Header -----------------------------------
   @NAME       : minctracc.c
   @COPYRIGHT  :
              Copyright 1993 Louis Collins, McConnell Brain Imaging Centre, 
              Montreal Neurological Institute, McGill University.
              Permission to use, copy, modify, and distribute this
              software and its documentation for any purpose and without
              fee is hereby granted, provided that the above copyright
              notice appear in all copies.  The author and McGill University
              make no representations about the suitability of this
              software for any purpose.  It is provided "as is" without
              express or implied warranty.

   @CREATED    : February 3, 1992 - louis collins
   @MODIFIED   : $Log: minctracc.c,v $
   @MODIFIED   : Revision 96.24  2011-02-24 20:02:35  louis
   @MODIFIED   : update for normalized mutual informatoin
   @MODIFIED   :
   @MODIFIED   : Revision 96.23  2010-04-01 04:49:16  rotor
   @MODIFIED   :  * fixed time.h include
   @MODIFIED   :
   @MODIFIED   : Revision 96.22  2009/05/26 18:03:07  claude
   @MODIFIED   : free more memory after usage
   @MODIFIED   :
   @MODIFIED   : Revision 96.21  2009/05/22 15:49:19  claude
   @MODIFIED   : fixed memory bug freeing initial transform
   @MODIFIED   :
   @MODIFIED   : Revision 96.20  2009/04/03 18:36:59  louis
   @MODIFIED   : made changes to use only DOUBLES for input source and model volumes, and for all estimation of deformation fields
   @MODIFIED   :
   @MODIFIED   : Revision 96.19  2009/03/13 19:51:31  claude
   @MODIFIED   : fixed bug in offsets for minctracc and free memory upon exit
   @MODIFIED   :
   @MODIFIED   : Revision 96.18  2008/10/08 15:17:49  louis
   @MODIFIED   : added -nmi option for linear normalized mutual information
   @MODIFIED   :
   @MODIFIED   : Revision 96.17  2006/11/30 09:07:31  rotor
   @MODIFIED   :  * many more changes for clean minc 2.0 build
   @MODIFIED   :
   @MODIFIED   : Revision 96.15  2006/06/04 07:02:35  rotor
   @MODIFIED   :  * Fixed 64 bit function pointer and ParseArgv problem with an enum for
   @MODIFIED   :       objective function type and interpolation type (thanks jason)
   @MODIFIED   :
   @MODIFIED   : Revision 96.14  2005/07/20 20:45:48  rotor
   @MODIFIED   :     * Complete rewrite of the autoconf stuff (configure.in -> configure.am)
   @MODIFIED   :     * Many changes to includes of files (float.h, limits.h, etc)
   @MODIFIED   :     * Removed old VOLUME_IO cruft #defines
   @MODIFIED   :     * Fixed up all Makefile.am's in subdirs
   @MODIFIED   :     * Removed all things in Proglib that are now part of MINC proper
   @MODIFIED   :     * Still working on fixing up perl subdirectory - removing mni_perllib
   @MODIFIED   :
   @MODIFIED   : Revision 96.13  2005/06/28 18:56:14  rotor
   @MODIFIED   :  * added masking for feature volumes (irina and patricia)
   @MODIFIED   :
   @MODIFIED   : Revision 96.12  2004/03/18 06:51:03  rotor
   @MODIFIED   :  * changed make_model from csh to sh
   @MODIFIED   :  * removed an extraneous printf from minctracc
   @MODIFIED   :
   @MODIFIED   : Revision 96.11  2004/02/12 05:54:22  rotor
   @MODIFIED   :  * removed public/private defs
   @MODIFIED   :
   @MODIFIED   : Revision 96.10  2004/02/04 20:42:33  lenezet
   @MODIFIED   : *** empty log message ***
   @MODIFIED   :
   @MODIFIED   : Revision 96.9  2003/02/04 06:08:44  stever
   @MODIFIED   : Add support for correlation coefficient and sum-of-squared difference.
   @MODIFIED   :
   @MODIFIED   : Revision 96.8  2002/12/13 21:16:11  lenezet
   @MODIFIED   : nonlinear in 2D has changed. The option -2D-non-lin is no more necessary. The grid transform has been adapted to feet on the target volume whatever is size. The Optimization is done on the dimensions for which "count" is greater than 1.
   @MODIFIED   :
   @MODIFIED   : Revision 96.7  2002/11/20 21:38:31  lenezet
   @MODIFIED   :
   @MODIFIED   : Fix the code to take in consideration the direction cosines especially in the grid transform.
   @MODIFIED   : Add an option to choose the maximum expected deformation magnitude.
   @MODIFIED   :
   @MODIFIED   : Revision 96.6  2002/08/14 19:54:42  lenezet
   @MODIFIED   :  quaternion option added for the rotation
   @MODIFIED   :
   @MODIFIED   : Revision 96.5  2002/03/26 14:15:38  stever
   @MODIFIED   : Update includes to <volume_io/foo.h> style.
   @MODIFIED   :
   @MODIFIED   : Revision 96.4  2002/03/07 19:08:34  louis
   @MODIFIED   : Added -lattice_diameter as an optionto minctracc to account for a
   @MODIFIED   : problem with the automated calculation of the sub-lattice diameter.
   @MODIFIED   : It used to be step*3*2 - which was pretty big, when step = 8mm.
   @MODIFIED   :
   @MODIFIED   : Now, the sub lattice diameter can be input on the command line, and I
   @MODIFIED   : suggest a lattice size 3 times greater than the step size.
   @MODIFIED   :
   @MODIFIED   : If not on the command line, the default is = 24mm.
   @MODIFIED   :
   @MODIFIED   : Revision 96.3  2000/03/15 08:42:41  stever
   @MODIFIED   : Code cleanup: all functions prototyped (except ParseArgs.c), no useless declarations, etc
   @MODIFIED   :
   @MODIFIED   : Revision 96.2  2000/02/20 04:01:03  stever
   @MODIFIED   : * use new history_string() function to generate history strings
   @MODIFIED   :   when outputting MNI files (.mnc, .xfm)
   @MODIFIED   : * removed unused vax routines from Proglib
   @MODIFIED   : * tuned configure script; CPPFLAGS and LDFLAGS are now left alone,
   @MODIFIED   :   for the installer to use
   @MODIFIED   :
   @MODIFIED   : Revision 96.1  1999/10/25 19:52:19  louis
   @MODIFIED   : final checkin before switch to CVS
   @MODIFIED   :
 * Revision 96.0  1996/08/21  18:21:51  louis
 * Release of MNI_AutoReg version 0.96
 *
 * Revision 9.6  1996/08/21  18:21:49  louis
 * Pre-release
 *
 * Revision 9.5  1996/08/12  14:15:42  louis
 * Never released version 0.95
 *
 * Revision 1.16  1996/08/12  14:15:40  louis
 * Pre-release
 *
 * Revision 1.15  1995/09/28  13:24:26  collins
 * added multiple feature volume loading with get_feature_volumes()
 * to be able to correlate multiple features at the same time.
 *
 * Revision 1.14  1995/09/28  11:54:52  collins
 * working version, just prior to release 0.9 of mni_autoreg
 *
 * Revision 1.13  1995/02/22  08:56:06  collins
 * Montreal Neurological Institute version.
 * compiled and working on SGI.  this is before any changes for SPARC/
 * Solaris.
 *
 * Revision 1.12  94/05/28  16:18:54  louis
 * working version before modification of non-linear optimiation
 * 
 * Revision 1.11  94/04/26  12:54:30  louis
 * updated with new versions of make_rots, extract2_parameters_from_matrix 
 * that include proper interpretation of skew.
 * 
 * Revision 1.10  94/04/06  11:48:43  louis
 * working linted version of linear + non-linear registration based on Lvv
 * operator working in 3D
 * 
 * Revision 1.9  94/02/21  16:35:51  louis
 * version before feb 22 changes
 * 
 * Revision 1.8  93/11/15  16:27:06  louis
 * working version, with new library, with RCS revision stuff,
 * before deformations included
 * 
 * Revision 1.7  93/11/15  13:12:10  louis
 * working version, before deform installation
 * 

Thu May 20 11:21:22 EST 1993 lc
             complete re-write to use minc library

Wed May 26 13:05:44 EST 1993 lc

        complete re-write of fit_vol to:
        - use minc library (libminc.a) to read and write .mnc files
        - use libmni.a
        - use dave's volume structures.
        - read and write .xfm files, for the transformations
        - allow different interpolation schemes on voxel values
        - allow different objective functions: xcorr, ssc, var ratio
        - allow mask volumes on both source and model
        - calculate the bounding box of both source and model,
          to map samples from smallest volume into larger one (for speed)

---------------------------------------------------------------------------- */

#ifndef lint
static char minctracc_rcsid[]="$Header: /private-cvsroot/registration/mni_autoreg/minctracc/Main/minctracc.c,v 96.24 2011-02-24 20:02:35 louis Exp $";
#endif

#include <config.h>
#include <float.h>
#include <ParseArgv.h>
#include <volume_io.h>
#include <minctracc.h>
#include <globals.h>
#include <objectives.h>
#include "local_macros.h"

/* objective function for nonlinear optimization.
 * Set in get_nonlinear_objective().
 */
static int obj_func0 = -1;

VIO_Real initial_corr, final_corr;
static char *default_dim_names[VIO_N_DIMENSIONS] = 
    { MIzspace, MIyspace, MIxspace };

/* Why are these declared here?  They aren't apparently used.
 * -smr
 *
 * static   const VIO_STR      TRANSFORM_FILE_HEADER = "MNI VIO_Transform File";
 * static   const VIO_STR      LINEAR_TYPE = "Linear";
 * static   const VIO_STR      TYPE_STRING = "Transform_Type";
 * static   const VIO_STR      LINEAR_TRANSFORM_STRING = "Linear_Transform";
 * static   const VIO_STR      GRID_TRANSFORM_STRING = "Grid_Transform";
 * static   const VIO_STR      DISPLACEMENT_VOLUME = "Displacement_Volume";
 */

/*************************************************************************/
int main ( int argc, char* argv[] )
{
	return minctracc(argc,argv);

}

